### ISSUE TITLE
Compatibility with Loom(virtual threads)

### ISSUE THREAD

**Issue by sghpjuikit** [haraldk/TwelveMonkeys#706]
JDK 19 has virtual threads and I was wondering if they can be used with imageio.

#### Why
To explain: loading images requires a lot of blocking io that uses threads very ineffectively (does not release them back to thread pool/os while waiting), but also lots of CPU for decoding or scaling. Neither of the two is inherently problematic, the problem becomes when these are together, intertwined. UI application uses event loop on main thread and is sensitive to 1 blocking the main thread or 2 CPU overload, either not giving UI enough time to run every even loop cycle, resulting in stutter. The former can be avoided by offloading blocking work to non main threads. The latter can only be fixed by not overloading CPU. Now that is much harder to do than it sounds!

The way to limit work done is using thread pool, say use at most 8 threads for image loading. This of course needs to account the physical nature of the CPU, i.e., the maximum number of threads busy with CPU work can be the number of cores the CPU has. Lets call this CPU thread pool. The problem is if the threads also do blocking io - they are not dong anything during that time - if during image loading 90% of the time is used for waiting, these 8 threads will only load N images with 10% of the speed! Obviously this is major problem. The solution is simple - just like with UI thread - never call blocking code on thread in CPU thread pool, by offloading all waiting (and waiting only) to another thread pool, lets call it IO thread pool. So IO would only run threads for waiting, these consume memory, but no CPU, so this pool has unlimited size and is there only so the CPU pool runs only CPU code so it saturates CPU with 100% usage. To make sure UI runs properly and to take other running applications into considerations, let say we initialize the CPU thread pool with smaller size than CPU core count, but lets not get into details.

To offload CPU threads from waiting, every blocking call is executed in IO pool. Ignoring the practicality issue of wrapping every blocking call into IO thread, even worse, if the blocking call executes some CPU code besides waiting, we are saturating CPU with CPU work from (an unlimited-size!!) IO pool, which disturbs the semantics of CPU pool for CPU and IO pool for waiting, causing CPU overload. Hence ONLY blocking calls must happen on IO pool. However if the blocking method that is called comes from library (almost certain), there is nothing one can do about it.

The result is that no matter what, one needs to assume how much workload is on CPU and how much is on IO and carefully finetune whatever thread pools one uses to sizes to achieve just the right CPU utilization. Anything less and user will have to wait until underutilized CPU finishes its work, anything more and user system comes to a crawling halt. The finetuning is impossible, because it depends on the exact CPU and exact workload.

If only there was something that could automatically delegate blocking to some void abyss and let threads do only CPU work. Well, virtual threads do exactly that. Every blocking call releases the thread and continues (on other thread, restoring/switching entire context to it) when blocking is done. During blocking no thread is used at all. This is the IO thread pool idea - except there is no IO pool, it is done for EVRY and ONLY blocking calls and its automatic. Thus we can block on CPU thread and still get 100% utilization without any thread pool size management!

#### Coroutines

Kotlin has something similar called coroutines. I wont go into details, what matters is that it has the limitation that if the blocking call mixes CPU and waiting inside library method, nothing can be done.

#### Usage

My use case is loading many images at once, potentially 1000s images on screen. These must load as quickly as possible, reach 100% CPU utilization but never overload CPU. This operation is sensitive - any less utilization and user waits a long time (just 1ms speedup per image means 1s of user time!. But the UI thread is already busy drawing all those images, so even slight more CPU utilization and the UI stars to stutter, resulting in horrible UX when user scrolls the view with images.

Right now I simply use the IO pool with size 64 (practically unlimited) to load images. However, CPU overload isnt nearly as dramatic as it would seem, but there IS stutter. I could play around the pool size, and I do have more tricks in my bag, but let's say I'm trying to make this simple and efficient. Using coroutines would not solve anything. Project Loom however could do wonders. SImply load images on CPU thread pool backed by virtual threads and everything should be perfect - in theory.

#### Loom issues

Loom is just out of the gate, and obviously not even in its final form.

Loom has a memory issue with ThreadLocal (many virtual threads cause unnecessary data instantiation).

Even worse - it does not support synchronization. This may be improved, but it is recommended to use Locks instead.

Also, the thread suspension for blocking calls is not really automatic - the blocking code has to be written with this in mind. Fortunately, even libraries rarely do its own blocking calls and mostly only rely on those in the JDK, which are already all retrofitted to support Loom. However, this concerns libraries that bring their own Input/OutputStream implementations and some more stuff:

From https://jdk.java.net/19/release-notes

```
The Mark and Reset Methods of InputStream and FilterInputStream Are no Longer Synchronized ([JDK-8284930](https://bugs.openjdk.org/browse/JDK-8284930))
core-libs/java.nio
The synchronized keyword is removed from the mark and reset methods of java.io.InputStream and java.io.FilterInputStream. This keyword serves no purpose as the other methods in these classes do not synchronize.

FileChannel.lock/tryLock Changed to Treat Size 0 to Mean the Locked Region Goes to end of File ([JDK-5041655](https://bugs.openjdk.org/browse/JDK-5041655))
core-libs/java.nio
The method java.nio.channels.FileChannel.lock(long position, long size, boolean shared) has been changed such that a size value of zero means to lock all bytes from the specified starting position to the end of the file, regardless of whether the file is subsequently extended or truncated.

FileChannel.transferFrom May Transfer Fewer Bytes Than Expected ([JDK-8286763](https://bugs.openjdk.org/browse/JDK-8286763))
core-libs/java.nio
The performance of FileChannel.transferFrom() has been improved significantly on Linux kernel version 4.5 and later for the case where the method is used to transfer bytes from one FileChannel to another. This change adds to the preexisting set of scenarios in which the number of bytes actually transferred might be less than the number requested to be transferred. That is to say, the value returned by transferFrom() can be less than the value of the count parameter: a “short transfer.” This is permitted by the specification, but might impact broken code that ignores the returned count and assumes it is always equal to count.
```

#### Question

1. Is this project Loom-ready?
1. If not, is there a plan or an ETA?
1. If Loom support required backwards-incompatible changes, what is the route the project would take?
1. Does this project use synchronization, ThreadLocal or native code?
1. Do the streams provided by this library break the Loom behavior of their superclasses? 

#### Adendum

I'm going to give this a try and try to provide some data as to what happens with JDK19 - does the application start, do the images load on  virtual threads, what is the performance, etc.

**Comment by sghpjuikit** [comment]
I made JDK19 work with Gradle. I had to use specific nightly Gradle version `distributionUrl=https\://services.gradle.org/distributions-snapshots/gradle-7.6-20220920234319+0000-bin.zip`

I made my project compile on JDK19. Had to use `--enable-preview` javac argument and `--enable-preview` JVM argument to be able to use virtual threads.

I managed to use `Executors.newVirtualThreadPerTaskExecutor()` ExecutorService instead of my IO executor.

Results:
* Images load properly. I had no immediate errors or problems. I expected errors, but everything worked like before.
* The preliminary test of performance shows great performance improvements! There is much less UI stutter, single image loading speed and total parallel image loading speed is the same. Setting `jdk.virtualThreadScheduler.maxPoolSize` to value smaller than CPU core count to give UI breathing room completely removed all UI stutter with no noticeable speed degradation.

This is absolutely revolutionary! Not a single line has been writte, No async code has been used. Great CPU utilization, no UI stutter. This literally could not go better! I'm speechless.

**Comment by haraldk** [comment]
Sound like you have answered most of your questions already, that's great! 😀 

I'll try to answer anyway:

1. I haven't made any Loom-specific testing. But I also don't see any specific reason why it should not work out of the box. Other than some ancient Java 2D initialization issue workarounds, there's very little synchronization or locking in the library.
2. Nothing special is planned.
3. That's a very hypothetical question... It would depend on the nature of the changes, I guess. 
4. * Synchronization: Yes, definitively, as there is a lot of that in the JDK stream implementations, Java 2D code etc., that we rely on. 
    * I find one usage of `ThreadLocal` in the code base, but I belive this is mostly "dead" code, so it could be removed. 
    * Native code: Not directly, but the JPEG plugin delegates to the JDK JPEG plugin which is basically the native libJpeg
5. As far as I understand, no; all the streams we provide typically only delegate reading to subclasses or JDK implementations. We decorate streams to provide on-the-fly encoding/decoding, buffering etc. But nothing really low-level/native that I would assume is what would break this.

So basically, I have no idea and give no guarantees, but I think there's a good chance of success. 

-- 
Harald K

**Comment by sghpjuikit** [comment]
1. The reason Loom may not work properly out of the box is due to things I mentioned - the use of synchronized or native calls. Lots of libraries will have real problems due to this. For more info see https://openjdk.org/jeps/425. Fyi, using `-Djdk.tracePinnedThreads=full` can help identifying illegal blocking calls inside virtual threads. Ill play around with it of course.
2. OK
3. Libraries like Hibernate, PostgreSql Java drivers or SpringBoot are already facing this hypothetical question, hence, it is not very hypothetical.
4.
  - Synchronization in JDK streams have been completely replaced with Locks, but with gotchas with subclassing. See the below official documentation, but in short (I even checked the code to confirm), subclassing certain streams alters the behavior to pre-JDK19 behavior, introducing synchronization in a Loom-unfriendly way. It appears imageIo is not affected, but your word has more weight than mine. From the link above:

> The java.io package provides APIs for streams of bytes and characters. The implementations of these APIs are heavily synchronized and require changes to avoid pinning when they are used in virtual threads. To prevent pinning, the implementations now work as follows: BufferedInputStream, BufferedOutputStream, BufferedReader, BufferedWriter, PrintStream, and PrintWriter now use an explicit lock rather than a monitor when used directly. These classes synchronize as before when they are sub-classed. The stream decoders and encoders used by InputStreamReader and OutputStreamWriter now use the same lock as the enclosing InputStreamReader or OutputStreamWriter.

> Additionally, the initial sizes of the buffers used by BufferedOutputStream, BufferedWriter, and the stream encoder for OutputStreamWriter are now smaller so as to reduce memory usage when there are many streams or writers in the heap — as might arise if there are a million virtual threads, each with a buffered stream on a socket connection.
  - ThreadLocals are not forbidden, they just might lead to higher memory. I would not be concerned about them for imageio.
  - com.sun.imageio.plugins.jpeg.JPEGImageReader right? It uses some synchronized and native calls. I'd assume they would rewrite this to be Loom-ready, but perhaps not. I still do not completely understand the Loom way of dealing with native calls, but yes, this reader may have some problems.

5. Ok, just keep in mind that subclassing certain streams reverts their behavior

It's not really about guarantees, but I wanted to hear from the author of the library to see what the potential issues and potential improvements are. Going into future, ideally, the library would make sure to support virtual threads, but that may take some effort to actually give guarantees. Of course, I'll report any troubles I might have. This is relatively new thing even for me.

I'm afraid Loom is going to be a little like Jigsaw - libraries are not forced to adopt it and may see it as too much effort. With Loom however there is a big upside for some users so there will be considerable pressure from now on for authors to adopt this.

**Comment by sghpjuikit** [comment]
I guess the issue can be closed. People can keep posting their experiences with Loom and imageio here.

**Comment by haraldk** [comment]
Thanks for investigating and reporting your findings! 

I might have misunderstood your question 3... If you mean how to deal with backwards incompatible API changes in the library,  we'll just make changes we need to. That might require bumping a major version number, but otherwise not a huge problem. However, there is not much API provided by the library (it's mostly plugins adhering to a predefined API), so I don't see a lot of changes needed.

Anyway, if you feel like, keep doing tests like that these, and ideally work them into unit tests or propose fixes/PRs, should you find any showstoppers! Keep in mind that this is an open source library, run by volunteers in their spare time, so we'll take any help we can get. 😀

**Comment by sghpjuikit** [comment]
Yes I meant backward incompatible changes and how this library deals with that situation, say if code breaks on new JDK but the change breaks on earlier JDK - I suppose only standalone artifacts can help at that point.

I am in the process of adopting JDK19 into my project and leverage virtual threads for image loading. I'll make sure to write up any new findings. As for help, I cant promise anything, but when I figure something out, I'll try to contribute (I already did make my own imageio artifact before).

**Comment by sghpjuikit** [comment]
Report: So far `-Djdk.tracePinnedThreads=full` did not report anything for me, not even for JPEGReader

**Comment by sghpjuikit** [comment]
Report: Conclusion after several days of use: Loading images seems absolutely fine with virtual threads. I tested common formats (but not gif), webp as well as psd and krita. There have been NO issues so far.

The only question for me is about how the `jdk.virtualThreadScheduler.maxPoolSize` should be set - this is so far the only way to throttle CPU load with virtual threads (without having specialized thread pool size limitation made for the use case). I understand this option should probably be left untouched, particularly in general application, but I want to do more investigation about how limiting virtual thread parallelism affects parallel image loading and UX.
